### ISSUE TITLE
Ignore typing annotations

### ISSUE THREAD

**Issue by balsagoth** [microsoft/pyright#108]
Pyright should ignore lines with  `# type: ignore` as PEP 484: 
https://www.python.org/dev/peps/pep-0484/#compatibility-with-other-uses-of-function-annotations

<img width="320" alt="Screenshot 2019-05-03 18 32 26" src="https://user-images.githubusercontent.com/183991/57151970-81d09100-6dd2-11e9-902b-8c807b2aa29f.png">

**Comment by erictraut** [comment]
I'm reluctant to add this support. I wouldn't want anyone on my team to use "ignore" comments in our code base.

There are already other good ways to provide the type checker with appropriate hints. This includes the cast operator and assert statements.

I'm interested in understanding where you have used "ignore" in the past and why you see a need for it.

**Comment by balsagoth** [comment]
I understand and agree with you to some extent, but I hate to leave false positive alerts on the code.

The other side is that without the possibility to ignore there is no easy way to silence the alert. 
And most of the times there is a solution. But sometimes we might encounter some situations that are not so easy to solve. For example:

<img width="801" alt="Screenshot 2019-05-03 23 00 11" src="https://user-images.githubusercontent.com/183991/57165988-0d5d1880-6df9-11e9-878e-353048354af1.png">
<img width="847" alt="Screenshot 2019-05-03 23 00 23" src="https://user-images.githubusercontent.com/183991/57165990-0d5d1880-6df9-11e9-812c-25ca09f4347a.png">

**Comment by erictraut** [comment]
Ah interesting. The specific case above is a bug in the typeshed stub file. I'll update the fallback stub file that ships with pyright, and I'll work to get it fixed in the official typeshed repo.

As a workaround, you could have used a cast operator in this case. (Yes, it's ugly, but so is "# type: ignore").

```
# TODO - remove cast once bug in datetime typeshed stub file is fixed!
new_datetime = dt.datetime.combine(cast(dt.date, date), cast(dt.time, dt.time.min))
```

**Comment by erictraut** [comment]
I just published 1.0.23 which fixes the bug in the datetime typestub file. If you see other problems which appear to be bugs in the type checker or in the type stub files, please report them. 

I'm going to leave this issue open a bit longer to see if others have input. I'm also interested to hear if you have other examples of where you've found "type: ignore" useful.

**Comment by balsagoth** [comment]
Thanks for the above hint.

The fix only silence the second error if I change 
```python
    if sys.version_info >= (3, 6):
        @classmethod
        def combine(cls, date: _date, time: time, tzinfo: Optional[_tzinfo] = ...) -> datetime: ...
```
to
```python
    if sys.version_info >= (3, 6):
        @classmethod
        def combine(cls, date: _date, time: _time, tzinfo: Optional[_tzinfo] = ...) -> datetime: ...
```

I use both, pyright and mypy, and I see differences between both sometimes. I'm actually curious why you don't include typeshed as a git submodule. For this particular case, I even don't understand why with the same typeshed pyright complains and mypy not.

If I found any good example to use `type: ignore` I will post here, until that I will try to fix all errors on my codebase.

**Comment by erictraut** [comment]
Ah yes, I missed the `_time` reference in the `>= (3, 6)` conditional. Thanks for pointing that out. I'll fix that in the next version.

I don't currently include typeshed as a git submodule because:
1. I don't want to include all of the third-party typeshed stub files
2. I don't want to include all of the stdlib typeshed stub files that apply to Python versions before 3.0
I guess I could include typeshed as a git submodule and then do post-processing as part of pyright's build scripts to package only the portions of typeshed that I want to include in pyright. Something worth considering.

If you see other places where pyright's outputs differ from mypy's — especially if you think it's a bug, I'd appreciate it if you'd let me know.

**Comment by erictraut** [comment]
I've fixed the typeshed stub file in 1.0.24.

I'm going to close this original issue because I don't think adding "type: ignore" is the right approach. But please post any additional bugs that you find in either the type checker or in the typeshed stub files. Thanks!

**Comment by carljm** [comment]
One place where we've found "ignore" comments to be almost indispensable is in gradually adding type coverage to a large untyped codebase. Often there are design issues with the existing code that make it not strictly type safe, and yet the code is already working in production. We much prefer to add coverage and acknowledge the existing issue(s) with a type-ignore (thus protecting the remainder of the code and all future modifications at least) than not add coverage. Of course the ideal is to fix the design of the existing code to remove all type errors, but universally enforcing this requirement places prohibitive drag on the pace of adding coverage. Even simply in terms of incremental development workflow, it's often much better (for code review, and to minimize the size of each change in e.g. continuous-deployment) to have one diff simply adding type coverage to unchanged code (with ignores where needed), and follow-up diffs doing the refactors needed to remove the type errors, rather than requiring all the changes to land in production at once.

Perhaps pyright doesn't intend to support the use case of gradual typing of existing untyped codebases?

I agree that I wouldn't ever support the use of type-ignore in greenfield code written with typechecking already in place, and it would be nice to have a way to disable type-ignore globally for a greenfield project.

**Comment by erictraut** [comment]
I'm currently going through the process of gradually typing a large existing codebase. That's what I had in mind when I wrote pyright, so it's very much a supported use case!

I haven't found "ignore" comments to be at all needed for this process. The process I'm using is documented [here](https://github.com/microsoft/pyright/blob/master/docs/getting-started.md). 

Many of the type-checking features of pyright are disabled in my pyrightconfig.json file because they would generate tens of thousands of errors in our source base. I'm visiting source files a few at a time and adding the `# pyright: strict` comment within the file. That enables all of the type checking rules for that file only. I then add type annotations to the file without changing any of the file's logic. This sometimes requires me to add assert statements to indicate where (possibly bad) assumptions are made within the code. For example, there are many cases where our code assumes that a method will never return None even though it does in certain error conditions. In those cases, I add `assert result is not None` immediately after the method call (sometimes with a `TODO` comment indicating that proper error handling needs to be added). In a few rare cases, I've needed to resort to using `cast` to work around an issue, but these cases are rare. In no case have I found a need to use `ignore`. 

Perhaps you could provide a concrete example of where "ignore" was useful and other techniques didn't work?

**Comment by carljm** [comment]
We aren't currently using pyright, so any specific example I might give from our experience might be handled differently by pyright. But here's one: sometimes we have code that stashes additional ad-hoc attributes onto certain object instances at runtime. Obviously this has never been a great practice, but it wasn't that uncommon in pre-typechecked days. And the typechecker rightly complains about accessing attributes that aren't part of the object's type. So the options I'm aware of for working around this are:

1. Add an annotation for the attribute to the type, even though it is not always present on all instances. This is not good, as it broadens the surface error of unsoundness.

2. Convert all access sites to use `getattr(..., ..., None)`, plus a following assertion that the `getattr` didn't return `None`. This would quiet the typechecker, but it's a more significant code change than just adding an assertion alone (e.g. it will change the exception type raised by a failure to get the attribute.)

For a diff that aims to add type coverage with zero behavior change, I consider an ignore comment preferable to either of those options.

One other wrinkle that might be relevant is that our typechecker (Pyre) has a syntax for type-ignore comments that specify the class of type error to ignore, so it's not a blanket silencing of any type error on that line.

**Comment by erictraut** [comment]
Yeah, I've run into a few instances of ad-hoc dynamic modification in our code base, but thankfully it's rare. In those few cases, I've added a call to "cast" to cast away the type (and added a TODO comment with a bug ticket to go back and fix it!). A "cast" call admittedly affects runtime behavior, but it's arguably just as safe as adding a type annotation.

**Comment by untitaker** [comment]
Another usecase for ignoring typing errors is code that monkeypatches third-party code. pyright is sometimes convinced that a certain attribute that I am trying to patch has never been defined in the external library, and I don't know how to work around this using casts or assertions.

**Comment by erictraut** [comment]
Why are the attributes unknown? Are they not defined in the type stub file for that third-party library? Or are they private implementation details that are not exposed in the type stub? If they're private, then monkey patching them is very fragile because it could change when you update the library. If they're public attributes, then they should be defined in the type stub file.

You can cast "x" to "Any" in your example above if you want to make arbitrary changes to the object or class without the type checker complaining. This effectively tells pyright to ignore all type checking.

Monkey patching in general is a hacky practice. I realize it's sometimes necessary, but the monkey patch code should be isolated in its own file and _extremely_ well documented. Pyright allows ignoring all type errors in a specified file or subdirectory by using the "ignore" array in the pyrightconfig.json file.

**Comment by untitaker** [comment]
I will cast the module to `Any` then. The attributes are unknown in two situations for us:

* we are monkeypatching a proprietary module that only exists within the context of an AWS Lambda function
* the third-party module is there, but pyright doesn't understand it

For the latter case you cannot argue for different coding practices because it is code I just have to deal with.

**Comment by untitaker** [comment]
For context, I am considering running pyright in addition to mypy for [Sentry's Python SDK](https://github.com/getsentry/sentry-python). I use `type: ignore` in many more places than the above, and I believe casting objects to `Any` will silence more errors than silencing a specific error on a specific line.

**Comment by fabiencelier** [comment]
One usecase in tests : we want to check what happens if someone calls a function of our library with the wrong type. (I can use cast to Any though)

**Comment by erictraut** [comment]
I don't fully understand your use case. You should be able to check an input variable's type (using "is" or "isinstance") even if it has a type annotation. Regardless, you're correct that you can always use a cast to Any if you need to do hacky things with types. There's usually a better solution than a cast, but in rare cases I've found that it's necessary.

**Comment by ironfoot** [comment]
I'd like to add one more support for "ignore" feature.

I need to skip a specific pyright bug.
No piece of code can be bug free and that includes pyright.
And no matter how responsive the maintainers are, fixing bugs will take time.
Meanwhile, I've been using pyright as a requirement for opening PRs.
I don't necessarily want to make pyright entirely optional. Just skip the bug until it's fixed.

**Comment by erictraut** [comment]
Pyright already supports multiple different ways to ignore an issue. One is Python's built-in `cast` call, which effectively tells the type checker "I know better than you do about this type". The second is an `ignore` list which allows you to ignore errors and warnings on a per-file basis. The third is a comment that allows you to enable or disable specific checks on a per-file basis (e.g. "# pyright: reportUnusedImport=false"). And there are almost always other ways to work around specific issues temporarily or permanently without using the above mechanisms.

Incidentally, we use pyright as a requirement for opening PRs on a large Python project too. It hasn't been a problem.

**Comment by carljm** [comment]
> Yeah, I've run into a few instances of ad-hoc dynamic modification in our code base, but thankfully it's rare. In those few cases, I've added a call to "cast" to cast away the type (and added a TODO comment with a bug ticket to go back and fix it!)

This is significantly inferior to a line-specific type-ignore comment, as you then lose all future typechecking on that variable on all later lines in what may be a large function.

**Comment by Emerentius** [comment]
Not supporting PEP 484 causes an incompatibility with mypy. Now, adding a `#type: ignore` comment on a line will cause *another* error from pyright, claiming that the type `ignore` doesn't exist. I've run into too many incompatibilities of pyright to abandon mypy for it (such as #114, sorry I haven't been keeping track of what else).

I don't find any of the alternatives you give particularly compelling replacements.
* `cast`: This affects runtime behaviour. If I typo that call or forget to use a forward reference for certain types, our service is now broken. Linters can help me find typos but I'm already drowning in error noise (very bad codebase). The linter delay (for all 3 of mypy, pylint, pyright) is too large to show those errors reliably before I may be off to another place in the code.
  `type: ignore` is exactly one of the tools I'm trying to address the flood of useless errors.
* pyright-specific file-based ignore features  
  Both the in-file comment syntax as well as the separate file based config have the same issue. They're overly broad. I don't want to whitelist entire files but only specific locations. Also, if I'm not mistaken, you can't disable type errors, is that right?

**Comment by untitaker** [comment]
Cast to `Any` works for sure but weakens the typing for me. `cast` is unacceptable due to runtime overhead.

**Comment by untitaker** [comment]
@Emerentius There is also the option to cast the expression to `Any`, which does not involve a call to `cast()`. I refuse to do this because it significantly weakens typing when all I want to do is to access a private attribute. But it is the option I would choose if I had to use pyright instead of mypy. It's the only one I know of that works in most cases and does not affect runtime.

In any case pyright is not fully PEP 484 compatible right now. That's okay, it's catering to different kinds of applications.

**Comment by erictraut** [comment]
I feel pretty strongly that "type: ignore" is the wrong approach. I wouldn't want anyone on my team using this in our code base.

That said, I recognize that others may have different goals in using a type checker. So as a compromise, I'm considering implementing support for "type: ignore" but providing a configuration switch to control whether it's enabled.

@untitaker, if you're aware of other places where pyright doesn't conform to PEP 484, I'd appreciate it if you'd file bug reports.

**Comment by erictraut** [comment]
I've added support for "# type: ignore" — both at the file level and on a per-line basis. I've also added support for a configuration switch "enableTypeIgnoreComments" that is on by default (for compliance with PEP 484) but can be disabled for those of us who don't want this construct to be used in our code bases. The functionality will be in the next release of pyright. Thanks to everyone on this thread for your feedback.

**Comment by untitaker** [comment]
Thanks for your work on pyright!

**Comment by erictraut** [comment]
This is now implemented in version 1.0.67, which I just published.

**Comment by gregdetre** [comment]
I'm having trouble getting the '# type: ignore' functionality to work. Here's my use-case.

I have defined an abstract class:

```
from abc import ABC, abstractmethod

class MyAbstractClass(ABC):
    @abstractmethod
    def __init__(): pass

class Concrete(MyAbstractClass):
  def __init__(): pass
```

I want to test it with Pytest:

```
def test_classes():
    Concrete()
    with raises(Exception):
        # should be abstract base class
        # pylint: disable=abstract-class-instantiated
        # type: ignore
        MyAbstractClass()
```

I've been able to silence Pylint, but I can't seem to silence the Pyright warning.

**Comment by gregdetre** [comment]
P.S. Pyright is super! I'm very grateful for it. Thank you.

**Comment by erictraut** [comment]
Pyright implements the "type: ignore" functionality according to the PEP 484 spec. It says: 
The # type: ignore comment should be put on the line that the error refers to:

A # type: ignore comment on a line by itself at the top of a file, before any docstrings, imports, or other executable code, silences all errors in the file. Blank lines and other comments, such as shebang lines and coding cookies, may precede the # type: ignore comment.

You've placed the comment on a line by itself but not at the top of the file, so pyright is ignoring it.

I've confirmed that if you place it on the same line as `MyAbstractClass()`, it suppresses the error.

**Comment by mangelozzi** [comment]
Howcomes it is not silencing the error:
![image](https://user-images.githubusercontent.com/19764314/131263096-0159143c-6d62-4344-ad97-bad0eb54d7e3.png)
The reason it is not accessed is because it an overriden dunction.

**Comment by erictraut** [comment]
The `# type: ignore` comment silences real diagnostics — errors and warnings. The message you're seeing here is a hint, but your editor is displaying it as a real diagnostic. I'd consider this a bug in your editor. Hints are not meant to be displayed in this manner. VS Code, by comparison, displays hints as a subtle "grayed-out" text effect, and it doesn't show the text associated with the hint.

**Comment by mangelozzi** [comment]
> The `# type: ignore` comment silences real diagnostics — errors and warnings. The message you're seeing here is a hint, but your editor is displaying it as a real diagnostic. I'd consider this a bug in your editor. Hints are not meant to be displayed in this manner. VS Code, by comparison, displays hints as a subtle "grayed-out" text effect, and it doesn't show the text associated with the hint.
> 
> Which editor are you using? You might want to report this as a bug to the maintainers of the editor or the LSP plugin for that editor.

Thanks for that information. I am using Neovim. The editor does show it as information, not red (error) or yellow (warning). I just changed the colors to be high contrast so I can read it easily. But its not the colour that is the problem, its about when going through a the list of diagnostic issues, that this one is flagged (and I have many other similar ones due to overriding/interfaces). Is there a way to say ignore this line completely (including hints)?

**Comment by erictraut** [comment]
Pyright reports these as information diagnostics with a "hint" tag. It reports these hints only when the client indicates that it supports them, so clearly the Neovim LSP is indicating that it wants to receive these hints, but it's not handling them correctly. These are not supposed to be interpreted as normal diagnostics. You should report the issue to the maintainer of the Neovim LSP. There's not really anything pyright can do to address this.
